torch/csrc/distributed/c10d/NCCLUtils.hpp

```diff
@@ -247,23 +261,40 @@ class NCCLComm {
       int numRanks,
       int rank,
       ncclUniqueId commId,
-      ncclConfig_t& config) {
+      ncclConfig_t& config,
+      bool eagerMode = false) {
     auto comm = std::make_shared<NCCLComm>();
+    bool isInitialized = false;
     if (nccl_use_nonblocking()) {
       config.blocking = 0;
-      C10D_NCCL_CHECK_TIMEOUT(
-          ncclCommInitRankConfig(
-              &(comm->ncclComm_), numRanks, commId, rank, &config),
-          comm->ncclComm_,
-          c10::nullopt);
+      if (eagerMode) {
```

Comment:
this is getting gnarly with the nested conditionals.  could it be restructured something like `if (eagerMode && nonblocking) { ... } else { ... }` to at least coalesce the two non-blocking inits into one?

Reply:
after some thought,  we don't really need to differentiate if eagerMode is enabled for non-blocking mode. As long as nonblocking mode is enabled, we always init comms asynchronously. and it is each collective's duty to make sure its comm is initialized. So we don't need 4 logical branches

Reply:
We can decouple the two conditions:
```
if (nccl_use_nonblocking()) {
  config.blocking = 0;
}

if (eagerMode) {
  C10D_NCCL_CHECK_NONBLOCKING(
    ncclCommInitRankConfig(...)
  )
} else {
  C10D_NCCL_CHECK_TIMEOUT(
    ncclCommInitRankConfig(...)
  )
  isInitialized = true;
}
```
That is, it should be safe to use `C10D_NCCL_CHECK_TIMEOUT` even if blocking mode is used (it will not enter the while loop in the macro).